include/seastar/util/log-impl.hh

```diff
@@ -53,37 +53,32 @@ public:
     class inserter_iterator {
     public:
         using iterator_category = std::output_iterator_tag;
-        using difference_type = std::ptrdiff_t;
-        using value_type = char;
-        using pointer = char*;
-        using reference = char&;
+        using difference_type = void;
+        using value_type = void;
+        using pointer = void;
+        using reference = void;
 
     private:
         log_buf* _buf;
-        char* _current;
 
     public:
-        explicit inserter_iterator(log_buf& buf) noexcept : _buf(&buf), _current(_buf->_current) { }
-        inserter_iterator(const inserter_iterator& o) noexcept : _buf(o._buf), _current(o._current) { }
+        explicit inserter_iterator(log_buf& buf) noexcept : _buf(&buf) { }
 
-        reference operator*() {
-            if (__builtin_expect(_current == _buf->_end, false)) {
+        inserter_iterator& operator=(char c) {
+            if (__builtin_expect(_buf->_current == _buf->_end, false)) {
                 _buf->realloc_buffer();
-                _current = _buf->_current;
             }
-            return *_current;
+            *_buf->_current++ = c;
+            return *this;
+        }
+        inserter_iterator& operator*() {
+            return *this;
         }
         inserter_iterator& operator++() noexcept {
-            if (__builtin_expect(_current == _buf->_current, true)) {
-                ++_buf->_current;
-            }
-            ++_current;
             return *this;
```

Comment:
Making the increment operators no-ops doesn't make sense for an iterator.
I'm not sure if there's a formal requirement for insert_iterators, but that would confuse any user
that is not familiar with the implementation internals.

Reply:
@michoecho  made the correct point that https://en.cppreference.com/w/cpp/iterator/back_insert_iterator does exactly the same.

I agree with you that I don't understand *why* though. Does it make for faster code? Simpler code? Something else? Are these reasons important this our log implementations?

I agree with @michoecho that his implementation should be good enough for any reasonable application... After all, this is a write-only iterator. What's the point of incrementing such an iterator without assigning a value? What is it supposed to do? Leave an uninitialized value? So if nobody ever wants to do that, making operator++ a no-op is good enough.